src/algorithms/synthesis/encoding.cpp

```diff
@@ -0,0 +1,126 @@
+#include "algorithms/synthesis/encoding.hpp"
+
+namespace syrec {
+
+    auto extend(TruthTable& tt) -> void {
+        // ensure that the resulting complete table can be stored in the cube map (at most 63 inputs, probably less in practice)
+        assert(tt.cubeMap.empty() || (tt.nInputs() <= std::log2(tt.cubeMap.max_size()) && tt.nInputs() <= 63U));
+
+        if (!tt.cubeMap.empty() && (tt.nInputs() > static_cast<std::size_t>(std::log2(tt.cubeMap.max_size())) || tt.nInputs() > 63U))
+            throw std::invalid_argument("Overflow!, Number of inputs more than 64");
+
+        TruthTable::CubeMap newCubeMap{};
+
+        for (auto const& [input, output]: tt.cubeMap) {
+            // ensure that all outputs are complete
+            assert(std::none_of(output.cbegin(), output.cend(), [](auto const& v) { return !v.has_value(); }));
+
+            // compute the complete cubes for the input
+            auto completeInputs = input.completeCubes();
+            // move all the complete cubes to the new cube map
+            for (auto& completeInput: completeInputs) {
+                newCubeMap.try_emplace(std::move(completeInput), output);
+            }
+        }
+        // swap the new cube map with the old one
+        tt.cubeMap = std::move(newCubeMap);
+
+        // construct output cube
+        const auto output = TruthTable::Cube(tt.nOutputs(), false);
+
+        std::uint64_t pos = 0U;
+        for (const auto& [input, _]: tt.cubeMap) {
+            // fill in all the missing inputs
+            const auto number = input.toInteger();
+            for (std::uint64_t i = pos; i < number; ++i) {
+                tt.cubeMap[TruthTable::Cube::fromInteger(i, tt.nInputs())] = output;
+            }
+            pos = number + 1U;
+        }
+        // fill in the remaining missing inputs (if any)
+        const std::uint64_t max = 1ULL << tt.nInputs();
+        for (std::uint64_t i = pos; i < max; ++i) {
+            tt.cubeMap[TruthTable::Cube::fromInteger(i, tt.nInputs())] = output;
+        }
+    }
+
+    auto encodeHuffman(TruthTable& tt) -> void {
+        std::map<TruthTable::Cube, std::size_t> outputFreq;
+        for (const auto& [input, output]: tt.cubeMap) {
+            outputFreq[output]++;
+        }
+
+        // if the truth table function is already reversible, no encoding is necessary
+        if (outputFreq.size() == tt.cubeMap.size()) {
+            return;
```

Comment:
this line should be easy to cover with a test that plugs in a reversible TT and asserts that the resulting truth table is the same.

Reply:
done 👍🏼